### ISSUE TITLE
Assembler and disassembler for Intel 8048

### ISSUE THREAD

**Issue by larsbrinkhoff** [PDP-10/its#1112]
MOON; 8478SA 7  
LMIO1; AS8748 38

Assembler and disassembler for the 8748 microcontroller found in CADR keyboards.  It's a variant of a 8048.

There's a 1K ROM image on Bitsavers.

**Comment by larsbrinkhoff** [comment]
Maybe Lisp machine code, but it looks like it might work in Maclisp too.

The assembly language program for the ROM is in LMIO1; UKBD 16

**Comment by larsbrinkhoff** [comment]
Photo my @MMcM.  I like how the PROM label has the ITS filename on it.

![](https://deskthority.net/resources/ribbon-cable-to-converter/22138)

**Comment by larsbrinkhoff** [comment]
AS8748 loads fine if LISPM is loaded first.

8478SA loads as is.

**Comment by larsbrinkhoff** [comment]
Also uses `multiple-value` and `return` with two arguments.  Maybe this is too LispM for Maclisp to handle.  I'm calling in an expert.  @eswenson1, what do you think?

**Comment by eswenson1** [comment]
Isn't that what the MLMAC provides?  (multiple-value-* functions, including multiple-value-return and multiple-value-list).  It should already be in DB.

**Comment by larsbrinkhoff** [comment]
Thanks, I see MLMAC defines a `multiple-value` macro.  I don't see that it overrides `return` so it accepts more than one argument.

**Comment by larsbrinkhoff** [comment]
I don't understand the problem with AS-ERROR.  It's defined as 

```
(defun as-error (error-string &rest args)
  (lexpr-funcall #'format t error-string args)
  (format t " while assembling ~A~%" as-form)
  (and as-known-new-pc (setq as-pc as-known-new-pc))
  (*throw 'AS-ERROR nil))
```

So if `&rest` works properly, this call shouldn't be an error:

```
(as-error "Phase error: Label ~A, Old PC=~O, New PC=~O" as-form (cdr label) as-pc)

**Comment by ams** [comment]
> Did zeta-lisp support multiple value return through a plain return special form? I don’t remember that. I do remember using multiple-value-return.


Yes, it is the recommended way of returning multiple values.

**Comment by ams** [comment]
> Did zeta-lisp support multiple value return through a plain return special form? I don’t remember that. I do remember using multiple-value-return.

Yes, it is the recommended way of returning multiple values.

MACLISP supports multiple return values via RETURN as well.

**Comment by ams** [comment]
I tried partial forms from this function on UP; and they all worked without loading any compatible issues.  I think the problem is elsewhere.

**Comment by larsbrinkhoff** [comment]
I tested putting the definition of `as-error` before `as-compile-form`.  That seems to fix the `Wrong number of args in function AS-COMPILE-FORM` error.

However, now I get

```
(COMMENT **ERROR**  ((LENGTH AS-FORM) 1.) 
                Computed function cant be numeric in function AS-COMPILE-FORM)
; DATA ERROR - TO PROCEED TYPE $P 
```

I think this is from `(≠ (length as-form) 1)` in `as-compile-form`.  Yes, that's a SAIL character.

**Comment by larsbrinkhoff** [comment]
I tested replacing `≠` with `neq`.  Now I get

```
(COMMENT **ERROR**  (RETURN ARG (QUOTE ADDRESS)) Wrong number of args in function AS-PARSE-ARG)
```

It doesn't help if I load `LIBLSP; LISPM` or `LISP; MLMAC` first.

**Comment by larsbrinkhoff** [comment]
I think the compiler has a point.  I tried to assemble the keyboard ROM in the Maclisp interpreter:

```
(load "liblsp; lispm")   ;Needed, or else the next load will complain.
T
(load "lmio1; as8748")
T
(load "lmio1; ukbd")
T

;Now run assembler
(as 'ukbd)
;((RETURN 1 ADDRESS) (NIL . 1)) TOO MANY ARGUMENTS - APPLY

;BKPT WRNG-NO-ARGS

**Comment by larsbrinkhoff** [comment]
@eswenson1, I'm not sure why this as8748 needs lispm manually loaded instead of autoloaded.  Or whether it should autoload or not.  But I think it's worth keeping in mind.

Or maybe add a compile-time load as mentioned in #1281.

**Comment by eswenson1** [comment]
You don't need manual loading -- you should use (eval-when (compile) (load '((liblsp) lispm))) (and include any other dependencies).  That's how it is "supposed to be done".  Autoloads are for when you want packages to automatically load in the interpreter on first reference.  Packages like SETF are MacLisp extensions that should autoload into the environment when first referenced.  However, it is not always the case that all the random dependencies used by packages like AS8748 should autoload.  This would raise them to first-class functions/variables and pollute the namespace.  I think it is fine for personal LISP startup files to add autoload properties as conveniences, or for "the system" to set up autoload properties for first-class, important functionality.  But autoloads are not the right solution for required compile-time dependencies.

**Comment by larsbrinkhoff** [comment]
I think I saw LISPM autoload in some other circumstance, so I thought it would be one of the "Maclisp integral" libraries.  But then, maybe not, and it was just loaded due to something else bringing it in.

**Comment by eswenson1** [comment]
Where are these files?  (MOON; 8478SA 7, etc)?  I don't appear to have them.  Can you post?

**Comment by larsbrinkhoff** [comment]
They are in http://github.com/PDP-10/its-vault.  files/moon/8478sa.7 and files/lmio1.

**Comment by eswenson1** [comment]
I'm seeing slightly different behavior.  I cannot compile moon;8478sa.7 I get errors because one of the functions defined in libdoc;loop > is not defined.  Are you sure it compiles cleanly for you?

**Comment by larsbrinkhoff** [comment]
I tested it just now, and no, it doesn't.

**Comment by ams** [comment]
Maybe this should be loaded on a LispM?

**Comment by eswenson1** [comment]
I'm sure it was intended to be run on a lisp machine.  But it should be able to be made to run under ITS MacLisp as well.  I don't see any real LM hardware/os dependencies in the code.

**Comment by eswenson1** [comment]
Ok, so UMLMAC in DB is busted.  I just recompiled it and now, after putting:

```
(eval-when (compile)
  (load '((lisp) umlmac)))
```

at the head of LMIO1; AS8748 >, I no longer get the error regarding SELECTQ (and the fact that LOOP-COLLECT-INIT is not defined in a UUO call).

I do, of course, get the phase error in AS-COMPILE-FORM that you reported earlier.  Will look into that next.  

NB:  UMLMAC defines SELECTQ in ITS MacLISP.  It is the "right" macro library to load to get SELECTQ in ITS MacLISP.  SELECTQ is in LM lisp natively.  It turns out there is some weird code in MLMAC regarding SELECTQ too, and it appears to just load the UMLMAC library and make sure that the version of SELECTQ defined there is used.  

Since AS8748 was written for ZetaLisp, it didn't need the (eval-when) form.  However, in ITS MacLISP, it does -- since SELECTQ is not part of native ITS MacLisp.

**Comment by eswenson1** [comment]
And the correct fix for the phase error is, as @larsbrinkhoff suggested, for the wrong-no-of-args error regarding as-error is to move the definition of as-error before first use.  This is because the compiler doesn't know the number of args that as-error takes -- and that number is variable. It remembers the number of arguments based on the first use of as-error (2 arguments).  When it encounters the 3-arg call (perfectly legal), it flags the error.  Moving as-error's definition first tells the compiler that this is an LSUBR (rather than an SUBR) and therefore n arguments are "ok".  

And, as you said, fixing this, leads to the "Computed function cant be numeric in function AS-COMPILE-FORM)" error.  Looking at that now.

**Comment by eswenson1** [comment]
@ams  On what basis are you saying that MacLisp supports multiple value returns using RETURN?  I don't see that documented in http://www.maclisp.info/pitmanual/contro.html#5.9.3, which I believe is the "latest" MacLisp documentation.  

From what I recall, you had to do (return (values x y z)) in MacLisp. 

In fact, looking at LSPMAI; LISP NEWS, I see this:

```
   1a) MULTIPLE-VALUE, in MacLISP, macroexpands into code to setq the requested
	variables to the multiple values returned by a function call.  E.g.
	    (MULTIPLE-VALUE (FORM EX?) (MACROEXPAND-1*M FORM))
	calls a macroexpander which returns two values -- the first is the
	possibly expanded form, and the second is a flag saying whether or
	not any expansion really happened.   If there aren't enough returned
	values to set all the indicated variables, the default action of the
	MacLISP facility is to pad out with ()'s (same as on LISPM); 
	however, **** if the global variable SI:CHECK-MULTIPLICITIES is 
	set to CERROR, an error will be signalled instead.   See the 
	discussion under item 4 below.
   1b) VALUES is a new LISPM special form, and in NIL/MacLISP it is 
	essentially the same thing as MSETQ-RETURN was; its basic action is
	to set up a "return-values vector";  typical usage would be to
	call VALUES as the last form in a lambda-expression.  Example:
	    (DEFUN TWIST (X Y) 
		(COND ((AND (NUMBERP X) (NUMBERP Y)) 
			(VALUES 'T (LESSP X Y)))
		      ('T (SETQ X (TO-STRING X) Y (TO-STRING Y))
			  (VALUES () (ALPHALESSP X Y)))))
   1c) MULTIPLE-VALUE-LIST constructs a list of all the (possibly multiple) 
        values returned by the evaluated sub-form.  Format of usage is like
	    (MULTIPLE-VALUE-LIST (<fun> ...))
	where <fun> should return some (possibly multiple) values.
   1d) RETURN in MacLISP still admits only one argument **** -- in order to do
	a RETURN with multiple arguments, do (RETURN (VALUES first second ...))
        MULTIPLE-VALUE-RETURN is primitive on the LISPM, but is macro expanded
	in MacLISP to use merely RETURN.  Currently, RETURN (but not
	MULTIPLE-VALUE-RETURN) is primitive in MacLISP, and this causes one
	incompatibility -- (RETURN (VALUES ...)) on the LISPM might return
	only one value, but in MacLISP it will return as many values as are 
	provided by the VALUES;  also, if FOO is a function returning
	multiple values, then (RETURN (FOO)) will "pass them all thru" doing
	essentially what MULTIPLE-VALUE-RETURN does on the LISPM.
   1e) VALUES-LIST is a SUBR of one argument, a list, which it "spreads out"
	into the "return-values vector";  Thus  (VALUES-LIST '(1 B 3))
	is the same as (VALUES 1 'B 3).
       RETURN-LIST macroexpands into something like 
	    (MULTIPLE-VALUE-RETURN (VALUES-LIST ...))
	As mentioned above for MULTIPLE-VALUE-RETURN, there is the implicit 
	LISPM/MacLISP incompatibility about RETURN.
   1f) MULTIPLE-VALUE-BIND -- a form is to be evaluated, returning several
	values, and *then* a lambda-binding contour is created, with
	the values bound to some variables;  following that, the body of
	and implicit PROGN is executed.  Example:
	    (MULTIPLE-VALUE-BIND  (A B C)  (GET-THREE-VALS ...) 
		(COGITATE-UPON A)
		(LIST B C))
```

**Comment by eswenson1** [comment]
Of course, if we want the code to still work on LISPM, then we should conditionalize the (eval-when ...) for MACLISP.  And probably write a macro to deal with turning the (return x y z ...) into (return (values x y z...)).

**Comment by eswenson1** [comment]
With the fix to UMLMAC, I can also compile MOON; 8478SA 7 without errors.  Shall I commit these changes (UMLMAC, AS748, and 8478SA)?  -- Eric

**Comment by larsbrinkhoff** [comment]
Sounds great.  No hurry.

> And probably write a macro to deal with turning the (return x y z ...) into (return (values x y z...))

Please do.  All those #+s are hurting my eyes! :-)

**Comment by eswenson1** [comment]
Ok, I wrote a couple macros to deal with the array handling (different in lispm/maclisp/nil), and now I get:

```
lisp^K

LISP 2154
Alloc? n


*
(load '((lmio1) as8748))
;Loading DEFMAX 98
132767
(load '((lmio1) ukbd))
T
(as 'ukbd)
;Loading MLSUB 17
UKBD
```

I assume that means it worked. Not sure how you are supposed to get a file out of this, but the UKBD symbol has the USER:LOCATION and ASSEMBLED-CODE properties on it, and the ASSEMBLED-CODE property looks like a bunch of assembled code.

**Comment by eswenson1** [comment]
I've pushed the required changes to the ejs/ticket_1112 branch.  Buiding now to see if UMLMAC gets built correctly.  I'll submit a PR if my build succeeds.

**Comment by larsbrinkhoff** [comment]
Thanks a lot!

We should also be able to run the disassembled code through Moon's disassembler.

**Comment by larsbrinkhoff** [comment]
Sorry, I mean the assembled code.  I'll edit.

**Comment by eswenson1** [comment]
I didn’t look very carefully, but how do you get the assembled code into the file system? Do you have to write your own code to take the code on the property list of the symbol UKBD and write to a file, or is there another utility?

Does the disassembler work with files?

I didn’t include UKBD PROM in my commit. Should I?

**Comment by larsbrinkhoff** [comment]
I don't know how the disassembler works.  I never got to the disassembler, since I didn't have any code to test with.

But let's do a quick peek...  I think it's takes an array of fixnums as input.

Yes, might as well include the PROM too.  It's related.

**Comment by larsbrinkhoff** [comment]
There's a UKBD PROM file, but it's text, not binary.

**Comment by eswenson1** [comment]
I'm inclined to submit a PR now, for this part.  The disassembler will also have to get ported to Maclisp -- it, too, uses LISPM-only constructs (array access).  Ok if I submit the PR for the assembler and source for disassembler, and then make a second commit/PR for the disassembler?

**Comment by larsbrinkhoff** [comment]
Sure!

**Comment by larsbrinkhoff** [comment]
But please add the original AS8478 38 and the updated 39 as two separate commits, and git rm the first from the second.

**Comment by eswenson1** [comment]
Uhoh.  I've already added them as one commit (with both files).  Not sure how to undo that now.

**Comment by eswenson1** [comment]
I guess I could just flush the whole branch, create another one, and do it the way you suggest (why do you want two separate commits for as8478 38 and as8478 39?  I currently have both files in the commit.  And I could wait until I get the disassembler ported too.

**Comment by larsbrinkhoff** [comment]
I can do the change to your branch, if you like.

The reason is that it makes it much easier to see the change, for reviewing now and also if someone else wants to check this in the future.

**Comment by eswenson1** [comment]
Ok.  Thanks!   And maybe let me know what you did so that eventually I can do this kind of surgery myself?  Thanks.

**Comment by larsbrinkhoff** [comment]
No problem, I kind of enjoy this kind of surgery.

**Comment by larsbrinkhoff** [comment]
You also have "Updated to build exthuk, gfile, gfn, and cnvd from source." on that branch.  That should stay, I guess?

**Comment by larsbrinkhoff** [comment]
I can split that off into a separate branch.

**Comment by larsbrinkhoff** [comment]
Now there's a ejs/exthuk with that first commit.

**Comment by eswenson1** [comment]
Now that you've done that (thanks).  I may hold off on a PR for this ticket until I update the disassembler to work on ITS MacLisp -- although it compiles, I see LISPM-only array use in there, so I'm sure it won't work.  I'll fix that first, try to prove that I can disassemble what was assembled, and then commit/PR.  Ok?

**Comment by larsbrinkhoff** [comment]
Updated ejs/ticket_1112, with forced push.

I did it like this:

0. First make sure everything is up to date.

   git fetch -p

1. Do an interactive rebase.   You'll see this in your editor:

   pick a628e70 Updated to build exthuk...
   pick 8fdd3f2 Added 8748 assembler...

   Change the first "pick" to "e".  This will let you edit the first commit.  But we'll not do that, but rather insert more commits in between the two.

   This is the trick.  We can use "git checkout ejs/ticket_1112" to pick apart the second commit.

2. Make a commit for the fixed UMLMAC.

   git checkout ejs/ticket_1112 -- src/lspsrc
   git checkout -p ejs/ticket_1112 -- build
   git commit -m umlmac

   Note the partial commit for "build".  We can now pick and choose changes made to that directory.  lisp.tcl as it happens.

3. Make a commit for the original AS8748 38.

   git checkout ejs/ticket_1112 -- Makefile src/lmio1/as8748.38 doc
   git commit -m original

4. Make a commit for fixing the assembler.

   git checkout ejs/ticket_1112 -- src/lmio1/as8748.39
   git rm src/lmio1/as8748.38
   git checkout -p ejs/ticket_1112 -- build
   git commit -m 'compile it'

5. Make a commit for the PROM.

   git checkout ejs/ticket_1112 -- src/lmio1/ukbd.*
   git checkout -p ejs/ticket_1112 -- build
   git commit -m assemble

6. Finally, proceed with the rebase.  There's just the disassembler left for the last commit.

   git rebase --continue
   Fix the conflict git complains about.
   git rm src/lmio1/as8748.38

**Comment by larsbrinkhoff** [comment]
Yes, your plan sounds very OK!

**Comment by eswenson1** [comment]
Thanks.  I’ll look over this.  I did a “git fetch origin” and then a “git reset –HARD origin/ejs/ticket_1112” and now, when I do a “git log”, I see this:

 

commit 2ace5fd52be5f4d644694cf88ba4753f52f1fc7d (HEAD -> ejs/ticket_1112, origin/ejs/ticket_1112)

Author: Eric Swenson <eric@swenson.org>

Date:   Fri Oct 5 09:13:23 2018 -0700

 

    Add 8748 disassembler.

 

commit 89a32904400b3417ba320b446753b41921138d2c

Author: Lars Brinkhoff <lars@nocrew.org>

Date:   Fri Oct 5 20:25:53 2018 +0200

 

    Use 8748 assembler to assemble Lispm keyboard PROM.

 

commit 236bfa6fd8154431f8a68d682d4a40f51b4c3644

Author: Lars Brinkhoff <lars@nocrew.org>

Date:   Fri Oct 5 20:19:12 2018 +0200

 

    Fix 8748 assembler to compile with Maclisp.

 

commit 69d16f48a0155fb39c8f2f4198710bfe110d5add

Author: Lars Brinkhoff <lars@nocrew.org>

Date:   Fri Oct 5 20:15:59 2018 +0200

 

    Add assembler for Intel 8748.

 

Why are there four commits?  Is this ok?  -- Eric

**Comment by eswenson1** [comment]
Never mind. I did a “git diff” of each delta and see that the last commit is the disassembler (just as the comment says).  I misread that and thought there were two commits where the assembler (and disassembler) were added.  Thanks for doing this.

 

From: Eric Swenson <eric@swenson.org>
Date: Friday, October 5, 2018 at 11:45 AM
To: PDP-10/its <reply@reply.github.com>, PDP-10/its <its@noreply.github.com>
Cc: Assign <assign@noreply.github.com>
Subject: Re: [PDP-10/its] Assembler and disassembler for Intel 8048 (#1112)

 

Thanks.  I’ll look over this.  I did a “git fetch origin” and then a “git reset –HARD origin/ejs/ticket_1112” and now, when I do a “git log”, I see this:

 

commit 2ace5fd52be5f4d644694cf88ba4753f52f1fc7d (HEAD -> ejs/ticket_1112, origin/ejs/ticket_1112)

Author: Eric Swenson <eric@swenson.org>

Date:   Fri Oct 5 09:13:23 2018 -0700

 

    Add 8748 disassembler.

 

commit 89a32904400b3417ba320b446753b41921138d2c

Author: Lars Brinkhoff <lars@nocrew.org>

Date:   Fri Oct 5 20:25:53 2018 +0200

 

    Use 8748 assembler to assemble Lispm keyboard PROM.

 

commit 236bfa6fd8154431f8a68d682d4a40f51b4c3644

Author: Lars Brinkhoff <lars@nocrew.org>

Date:   Fri Oct 5 20:19:12 2018 +0200

 

    Fix 8748 assembler to compile with Maclisp.

 

commit 69d16f48a0155fb39c8f2f4198710bfe110d5add

Author: Lars Brinkhoff <lars@nocrew.org>

Date:   Fri Oct 5 20:15:59 2018 +0200

 

    Add assembler for Intel 8748.

 

Why are there four commits?  Is this ok?  -- Eric

**Comment by larsbrinkhoff** [comment]
There should be five commits:

- 2ace5fd Add 8748 disassembler.
- 89a3290 Use 8748 assembler to assemble Lispm keyboard PROM.
- 236bfa6 Fix 8748 assembler to compile with Maclisp.
- 69d16f4 Add assembler for Intel 8748.
- 864dc86 Fix building UMLMAC.  Update herald version.

**Comment by eswenson1** [comment]
I'll push the changes to make the disassembler compile and not do LISPM-only things on Maclisp.  However, I don't know how to use this.  I've run FROB, which appears to load the data in a table into an array.  Then, I tried to run DISAR, but don't know what to pass it.  The first (array) argument is supposed to be the program to disassemble.  However, I don't know how to get it there.  It appears to need to be an array of fixnums, since the contents of the array are used to index into the array that was created from the table.  However, the output of the assembler is a property list on a symbol. One of the properties contains a list of what appears to be the assembled code, however, not all elements of the list are fixnums.  Some are lists (e.g. "(= 0)").  So although I could write some code to convert the list to an array (so that I could pass it to DISAR), some of those elements would not be fixnums and would lead to errors attempt to access the  disassembler's internal array.  So I'm not sure how one is supposed to prepare data for disassembly.  Any ideas?

**Comment by eswenson1** [comment]
Pushed my latest changes.  I'll enter a PR so that we can get this merged and more of us can look at it.

**Comment by larsbrinkhoff** [comment]
I don't have any more ideas at the moment.  I'll run the assembler and prod around when it's merged.

**Comment by ams** [comment]
> @ams On what basis are you saying that MacLisp supports multiple value returns using RETURN? I don't see that documented in http://www.maclisp.info/pitmanual/contro.html#5.9.3, which I believe is the "latest" MacLisp documentation.

It was a munge on my side, I only tried without explicitly calling return. :/
 
```
(defun foo () (values 1 2 3))
FOO
(foo)
;Loading DEFMAX 98
;Loading MLMAC 92
(multiple-value-list (foo))
;Loading MLSUB 17
(1 2 3)
```

Indeed, if you do RETURN with multiple arguments you will get a BKPT on WRNG-NO-ARGS.

**Comment by ams** [comment]
FWIW,

```

;FORMAT OF A PROM FILE IS:
; FIRST LINE IS THE LABEL TEXT.
; SECOND LINE IS BLANK
; THEN COME PROM LOCATIONS, ONE WORD PER LINE, IN THE FORM
;  ADDRESS, TAB, CONTENTS, SPACE, CRLF.  NUMBERS ARE IN OCTAL.
; AT THE END WE HAVE A BLANK LINE AND THE WORD END.

```

From MOON;PROM BURNER.  In later versions of the Lisp Machine system i cannot find a disassembler. :(

**Comment by larsbrinkhoff** [comment]
Maybe the binary data never had to go to a file.  MINITS; PROMP DOC has this:

```
    To program a prom or a pal you need to have the data stored in a
lisp array in ascending order and to have this array be the value of a
symbol.  To program a prom call:
     (PROGRAMMER-PROGRAM-PROM 'Symbol-with-array-value)

**Comment by larsbrinkhoff** [comment]
Fixed by #1314.

**Comment by ams** [comment]
https://github.com/LM-3/sys/issues/11
